### ISSUE TITLE
Using LanguageServerWrapper#getServerCapabilities() can be dangerous and provides a lot of freeze

### ISSUE THREAD

**Issue by angelozerr** [eclipse-lsp4e/lsp4e#628]
I create this issue because I'm porting LSP4E in IJ and the call of this method can freeze the IJ and I have an idea how to avoid doing that but I would like to know your feedback.

The main problem when you call LanguageServerWrapper#getServerCapabilities() is when:

 * you are closing the last file mapped with the language server which reports diagnostics and a component is calling LanguageServerWrapper#getServerCapabilities(). We have this usecase in IJ because we need to check the code action capability when we reports diagnostics. If we close  IJ and there are diagnostics which are reported when language server is stopping, we call LanguageServerWrapper#getServerCapabilities()  which restarts the language server which is bad. Perhaps in LSP4E there is this kind of usecase, I don't know.
 * the language server cannot be started and in this case there are a lot of freeze (because we wait for 10 seconds) but more we could wait for 2 * 10 seconds. For instance this code https://github.com/eclipse/lsp4e/blob/71fdc020de59658be6a29205f24d4f2c56ea2eaa/org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServiceAccessor.java#L300 will wait for 20 seconds because we call twice the getServerCapabilities.

My idea is to cache the serverCapabilities instance in a LanguageServerDefinition. The idea is to wait for 10 seconds just one time (if the language server cannot be started for instance) and after that we use the cached serverCapabilities. It will give the capability to check server capability even if the language server is not started the second time.

What do you think about this idea?

**Comment by mickaelistria** [comment]
I don't think that capabilities should belong to the server definition, they're a per-instance piece of data.
Before changling LSP4E, can you please provide steps to reproduce this issue? The fact that it fails in IJ isn't per se a strong enough reason to change LSP4E.

**Comment by angelozerr** [comment]
> I don't think that capabilities should belong to the server definition, they're a per-instance piece of data.

Why? All language server wrapper instance will return the same server capabilities for a given server definition, no?

> Before changling LSP4E, can you please provide steps to reproduce this issue? The fact that it fails in IJ isn't per se a strong enough reason to change LSP4E.

It is very difficult to provide a usecase which have some freeze, because getServerCapabilities can be consume in a lot of context.

**Comment by mickaelistria** [comment]
> All language server wrapper instance will return the same server capabilities for a given server definition, no?

That is not specified; and it's possible that some language server has different capability after a restart (eg an extension or 3rd party was installed and makes one feature now available; or the LS has a marker somewhere that will cause a feature to be disabled in further executions because it has caused major failure...)

**Comment by angelozerr** [comment]
Ok thanks @mickaelistria to have taken time to give me an answer.

I thinl https://github.com/eclipse/lsp4e/blob/71fdc020de59658be6a29205f24d4f2c56ea2eaa/org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServiceAccessor.java#L300 should be fixed, because getServerCapabilities is not a simple getter.

**Comment by vgcpge** [comment]
> > All language server wrapper instance will return the same server capabilities for a given server definition, no? 
> That is not specified; and it's possible that some language server has different capability after a restart (eg an extension or 3rd party was installed and makes one feature now available; or the LS has a marker somewhere that will cause a feature to be disabled in further executions because it has caused major failure...)
Also, the protocol specifies, that capabilities are established upon initialization. 

But, on the other hand, this does not preclude LSP4E from making capabilities eventually consistent, quickly returning  cached data associatedwith server definition and notifying clients upon updates.

Still needs the test steps though.

**Comment by sebthom** [comment]
@angelozerr can this issue be closed now that getServerCapabilitiesAsync() is available?
